ENH: linalg.solve: implement banded solver in cpp

BasilLiekens opened the pull request:
#### Reference issue
This PR contributes to #23774 by moving the last structure for `solve` into the lower level implementation.

#### What does this implement/fix?
The `gbsv` - `gbcon` sequence previously performed in `solve0` is now also implemented directly in the lower-level implementation, meaning all necessary calls can be made directly from `solve` and the need for `solve0` is removed.

#### Additional information
There are still a few questions left unanswered about this implementation, cfr. https://github.com/scipy/scipy/issues/23774#issuecomment-3630774818

Of these questions, two have been answered already:
1. `np.atleast_2d()` is not a ready-made substitution for the sequence laid out as it appears to append dimensions in the front rather than the back as would be necessary.
2. The comment about the fact that structure detection is commonly left out has been addressed.

BasilLiekens committed:
Implemented banded solver in cpp

- Moved the LAPACK routine 'gbsv' to two separate steps: factorization using 'gbtrf' and 'gbtrs' for consistency with other solvers used.
- Added other necessary machinery to make everything work: LAPACK overloads, norm computation, reshuflling routines etc.

ev-br commented on scipy/linalg/src/_linalg_solve.hh:
```diff
@@ -388,6 +424,8 @@ _solve(PyArrayObject* ap_Am, PyArrayObject *ap_b, T* ret_data, St structure, int
                 slice_structure = St::LOWER_TRIANGULAR;
                 uplo = 'L';
+            } else if (upper_band + lower_band < n) {
+                slice_structure = St::BANDED;
```
This is new relative to to main (which skips "banded" in the structure structure detection), and it goes against @ilayn's suggestion at https://github.com/scipy/scipy/issues/23774#issuecomment-3631006935

The rest of `solve` mainly follows Matlab's mldivide, cf the "dense inputs" block-scheme in https://www.mathworks.com/help/matlab/ref/double.mldivide.html which does not have "banded"

So this probably needs to be removed in the end. Meanwhile I'm curious what the performance characteristics are for few bands (e.g. N=100 and n_bands=5). So could you
- add  a benchmark to https://github.com/scipy/scipy/blob/main/benchmarks/benchmarks/linalg.py#L105
- run it with `$ spin bench --compare main -t BatchedSolveBench` and report the results here
- remove this in a separate commit to simplify further experimentation.

ev-br replied:
> Should the benchmark also be committed

Yes certainly.

ev-br reviewed (changes requested):
Initial cursory review, will do a more detailed sweep through later.

This looks like a great start @BasilLiekens . I left a several comments below. Also, we definitely need to run the CI, so please remove `[skip ci]` from the commit message. This also needs some more tests, especially large number of bands (what does it do when `n_upper = n_lower = n`?

BasilLiekens commented on scipy/linalg/src/_common_array_utils.hh:
```diff
@@ -622,6 +668,23 @@ void copy_slice_F_to_C(T* dst, const T* src, const npy_intp n, const npy_intp m)
+ * Compute the transpose of an m x n F-ordered array `src` inplace using scratch memory
+ */
+ template<typename T>
+ void transpose(T* src, T *scratch, CBLAS_INT m, CBLAS_INT n) {
```
This might indeed be a good idea: theoretically transposing and swapping C to F ordering should be identical. However, if a different set of routines would be used the need for this would be eliminated either way. (See other comment on the routines.)

BasilLiekens committed:
ENH: added basic benchmark for banded

ilayn commented:
Structure of an array or ndarray with banded slices is not the structure `assume_a` accepting functions expecting. The `assume_a` works on the common properties of all the slices. The point assume_a is to delegate the operation to a specialized solver, trying to reduce the constant term in `O(n^3)` linear system solution. And **not** to provide API convenience because each solver has multitude of different options and trying to harvest each type options makes things very complicated very quickly. Typical solution with `solver_options={}` etc. comes in but overall it becomes a UX mess especially for inline usage of `solve`. This is mostly because we are trying to serve both batched input convenience and solver specializations which are in conflict with another. There are also some historical luggage we are carrying too which is not helping. 

Here structure word is becoming confusing because they all have structure, they are banded afterall, but not exactly the same. That's why typically bandedness is left out from this catalogue because being bandedness do not lend itself to be treated exactly for each slice as you mentioned, `kl` and `ku` per slice might be different. 

So either we do a `bandwidth` run on each slice with a slight performance degradation for fresh `kl`, `ku` when the user selects `assume_a="banded"`, or we push this to `solve_banded` and handle things specialized to it. I think the former might be a good compromise since bandwidth has negligible cost compared to the solving stage.

BasilLiekens commented on scipy/linalg/src/_linalg_solve.hh:
```diff
@@ -449,6 +487,23 @@ _solve(PyArrayObject* ap_Am, PyArrayObject *ap_b, T* ret_data, St structure, int
+            case St::BANDED:
+            {
+                // guard against the case where `assume_a` was set directly, but now runs this routine twice.
+                bandwidth(data, n, n, &lower_band, &upper_band);
```
Actually, there is a problem with the guarding approach: since the `lower_band` and `upper_band` variables are set by the call to `bandwidth`, this guard would be ineffective for subsequent slices. 

Since it was confirmed that the `banded` option should be manually set, it might be a good idea to just leave it as is as there would be no redundant calls either way. 

The only problem that could arise with this would be that the memory allocation for `work` would not be able to happen based on the maximum required, but it might be best to just allocate `3n^2 + ...` flat out in that case since the allocation happens first. One could add a statement here https://github.com/scipy/scipy/blob/2325d7ce20a9b862d9e95039cec3f4f2353c3f21/scipy/linalg/src/_linalg_solve.hh#L345-L347, but then one would have to decide whether to keep track of the bandwidths of all slices (necessitating extra allocations), or to just stick with the max, which might be inefficient.

BasilLiekens commented:
> what does it do when `n_upper = n_lower = n`?

Some preliminary tests show that this can lead to segfaults. There is, however, not really a direct, clearly best solution to this problem: 
- One could just allocate an additional `3n^2 + n` elements in the `work` array, but this does not really feel like an elegant solution as it would lead to a very big chunk of memory remaining unused in a lot of cases, not that it would matter that much probably, but it doesn't really feel right.
- Another one would be to do the maximum bandwidth detection over all slices and based on that allocate an extra array, say `ab`, of the correct size, but this feels a bit like jumping through extra hoops, overcomplicating things. Especially since that memory allocation could fail, which would then require freeing the memory once again etc.


When it comes to the link with `solve_banded`, there is one complication I could see with the suggested approach: one would have to get all slices into the correct shape such that they all fit the "largest bandwidth" case, which again breaks a bit the common control flow that is followed now in the `solve` function of  `preliminaries (allocations etc.) => loop over slices => hand each slice of to correct routine`. 

Therefore, it might be easier to just follow the same processing flow for `solve(a, b, assume_a="banded")`, but then make the current `solve_slice_banded()` take care of the rearrangement and handing it of to the proposed internal `_solve_banded()`. In other words, let the proposed `_solve_banded()` also operate on individual slices. This would simultaneously allow to take into account the different bandwidths of different slices. Furthermore, this also allows for easy expansion of `solve_banded()` should the API one day change to also allow for array inputs for the bandwidths.

BasilLiekens commented:
All requested changes should have been implemented now:
- The `transpose` routine has been replaced by a combination of the already provisioned `swap_cf()` and `copy_slice()` routines. In either case one of the two routines is used to ensure the same number of copies. However, one could also argue that the `trans == 'T'` case is less frequently used, and that it is therefore acceptable to incur some form of performance penalty in that case.
- The size of the allocated array is decided on based on a preliminary sweep of the arrays to determine the bandwidths, resolving the issue of the segfaults in the case of full arrays that has been brought up. Some additional memory is provisioned for the copy/transpose operations from the previous point. If that were to be lumped somehow, the allocated amount of memory could be brought dow as the scratchpad memory could then be overwritten.
- The case of  `info < 0` is addressed and the "choice" of routines has been discussed in the previous comments, but TLDR, the reason is that it allows for an easier move of `solve_banded` into the C implementation.
- The automatic detection of the banded structure has been removed and a simple benchmark has been added.
- For the aforementioned memory allocation a precursory sweep of all slices is done to find that required amount of memory that should be allocated. (This amount could further be resolved as previously mentioned.)
- The `overwrite_b` typo has been fixed. Unfortunately, this also leads to a linting error as the value is not used. Surprisingly however, no error is raised for the `overwrite_a` parameter which is also not used. 

In the implementation some design decisions in the changes can be debatable (cfr. for example the transposition alongside some of the freeing techniques in the bandwidth sweep), so these can be changed if desired.

ev-br commented on scipy/linalg/src/_linalg_solve.hh:
```diff
@@ -276,6 +318,11 @@ _solve(PyArrayObject* ap_Am, PyArrayObject *ap_b, T* ret_data, St structure, int
+    // Only used for `banded`
+    npy_intp *kls = NULL;
+    npy_intp *kus = NULL;
+    T *ab = NULL;
```
ISTM it will be cleaner to have a separate `_solve`-like function just for the banded mode. Then you just allocate exactly what needs allocating, you know exactly the structure etc. The code duplication will be relatively mild; the code will be simpler and easier to reason about.

ev-br reviewed (commented):
I think it'll be cleaner and more maintainable to branch for `assume_a="banded"` earlier, somewhere around the `// Workspace computation and allocation` part of `_solve`. Then detect the bandwidth, allocate exactly what needs allocating for the banded routines, copy to the banded storage straight from a strided lhs array etc.--- all in a separate function.

ilayn commented:
Let me elaborate more

> When starting this port, the following code was used as a base: 

I did not know that we pushed the banded version in so that's my mistake missing that. So thank you for clarifying that. However there are some concerns that need to be addressed I guess. 

- So the back story is that we wanted to do this polyalgorithm originally for the 2D arrays only and only for solve until we settle the API. 
- Then things got a bit fast and the batching of linalg happened first. 
- Then we started pushing the batching to low level. 

That's why things are a bit in reverse as we batched first and now trying to figure out the low level API. This is shadowing a few things. 

solve_banded, lu_factor, and other functions are historically added as convenience functions over the LAPACK functions. And lived as they are without any relationship to solve. That's why solve_banded inputs are prepared to be passed to banded LAPACK routines. 

We are now trying to bridge between these functions. The reason might be code duplication, or serving two functions at the same time with the same function. But every time this is done, it is the next maintainers job to handle all that complexity/clean up. I have done this already a few times and there is no need to be this frugal about the lines of code. It only brings in unsolicited complexity. As I mentioned `lstsq` was tied to `pinv`, which was tied to `pinv2`. Similarly `eigh` was tied to `pinvh` at the same time and `eigvalsh`. And making them work better with still carrying all those relationships is just unnecessary work regardless of the small convenience it might bring in.

Once the solve mechanism is in, we can do `solve_banded` versions, presumably. However there is no need to create coupling between solve and any other function just yet because the API is still settling. These early couplings become historical warts on the next maintainer. So it might sound a good idea for minute gains but it is not justified for 100 lines of code. Moreover, `solve` is the most used function and `solve_banded` is a niche function that nobody really cares. Tying them up right now would be a mistake and inflates the concerns unnecessarily. I don't have any objections if we do it later but at this stage it is just complexity for complexity sake. This is what I mean about the `solve_banded` in case it was not clearer earlier. This is a hill I would like to die on since I have sifted through many examples of this pattern, and that's why it catches my eye. 

--------------

Regarding the switch to banded solver based on heuristics, we discussed this above, but just to gather all arguments together, there is no generally-applicable heuristic for all platforms/architectures and starts to annoy users that n=55 is t seconds, but n=56 seconds is something else. This has happened in many tools historically so people stopped doing it if the heuristic is not continuous in spent time. 

-------------

Regarding combining `trf` and `trs` to `sv`, is not due to performance reasons but to inject `gbcon` call in between so that we can emit warnings about ill-conditioning (run `linalg.solve(np.arange(9).reshape(3, 3)+1e-15, np.ones(3))`). Besides `sv` is just `trf`+`trs` calls sequentially. Hence there is not much difference. We are paying the `gbcon` call cost for the warning generation. Skipping the warning is making more surprises on the user side hence it is a common practice to include it despite the cost.

-------------

I am not quite sure why you are doing the slicing for `solve_banded`, since I thought we already do the slicing already in `solve` C++ function. I did not check it yet unfortunately what we have but I was under the impression that the following pseudo-code was the case

```Cpp
void
solve_cpp_function{PyArray_Object* arr, ....)
{
     // Array preparation size and other details

     /* ---------- MAIN SLICE LOOP ----------- */
    for (int i; i < number_of_slices; i++)
    {
         // if structure is given if not discover
         structure = ....

         // Branch out to the relevant solver
        if (structure == diagonal) {

        } else if (structure == banded) {

             // copy slice into banded format
             // make calls to gb routines
             _solve_banded(...)    // <- I was under the impression that you were implementing this

        } else if (structure == posdef) {
              // go to cholesky test etc.
        } else if (...) {
              ....
         }


         // Get the next slice
    }
     /* ---------- ----------------------- ----------- */
    // ... Clean up and return
}

```

So I thought, the slicing belongs to the main loop, so that we can mix and match different solvers in case in the future we decide to implement a novel heuristic. An example of this is used in matrix function algorithms;

https://github.com/scipy/scipy/blob/6ff4d18ff729b9fe6389d814390ac355d2c89f42/scipy/linalg/src/_matfuncs_sqrtm.c#L71-L101

But as I mentioned, I have not followed the C++ implementations closely hence, if it is not the plan then nevermind, I'll leave it to @ev-br.

ev-br commented:
>  if it is not the plan then nevermind, I'll leave it to @ev-br.

Yes, this is pretty much how the current implementation works. `banded` is not in it though, currently. Only `diagonal`, `tridiagonal`, `symmetric / hermitian`, `pos.def` and `general` are.

As long as we do not include `banded` into the structure detection, I think it's better to branch off earlier if `assume_a="banded", so that the logic is

```
if assume_a == "banded":
   return _solve_banded(...)
else:
   # loop over the slices, detect per-slice structure etc


def _solve_banded(...):
      # figure out the bandwidth
      for slice in slices:
            copy_bands
            call lapack
            ...
```

This way we are duplicating the looping over the slices, yes, but I agree it is small potatoes. What we get by this early branching is
- easier for maintenance and potential future evolution (if we want to shove it into the main loop, it is easy)
- easier to understand and follow
- better for memory pressure.

This is basically https://github.com/scipy/scipy/pull/24123#issuecomment-3637489112; we can leave the public `linalg.solve_banded` function off it --- I personally care about this function, but we can totally leave it to some future point in time.

BasilLiekens commented:
The `gbsv` routine has been swapped out for `gbtrf => gbtrs`, making everything more aligned in terms of actual calling conventions compared to the other cases. Furthermore, this also removes the need for manually handling the transpositions. Overall, this feels like a more ergonomic solution than the previous one.

> I am not quite sure why you are doing the slicing for solve_banded, since I thought we already do the slicing already in solve C++ function.
>> As long as we do not include banded into the structure detection, I think it's better to branch off earlier if `assume_a="banded", so that the logic is ...

That is basically what is done now as suggested in https://github.com/scipy/scipy/pull/24123#pullrequestreview-3613807620. This is mostly a consequence of the fact that some additional allocations are necessary for the `assume_a=banded` case, which would clutter the main `_solve()` function body. Should a heuristic to detect a banded structure ever be introduced it should be pretty easy to bring `_solve_banded()` back into the main loop.

--- 

This PR should be ready now, except for the `overwrite_a` and/or `overwrite_b` arguments; as mentioned in a number of previous comments, the `overwrite_b` argument typo that was fixed is now causing linting errors. The way I currently see it there are a few options to go about this:
- Leave everything as is, and leave the typo in. (This would fix the linting errors, but is not really a good solution as it will probably just remain in there until someone else runs into the same issue.)
- Add a ruff `# noqa: F841` directive on line 247 in the current implementation to suppress the warning. This fixes the typo without the linter complaining, but it does not fix the underlying issue.
- As mentioned in https://github.com/scipy/scipy/pull/24123#issuecomment-3694097250, the `overwrite_a` argument is not even passed into the lower level implementation. (The fact that it is unused is already documented in https://github.com/scipy/scipy/issues/23141.) Therefore, a third option would be to add an extra argument to the `_solve()` call that allows for passing in `overwrite_b` and then passing in both the `overwrite_x` arguments into the lower-level implementations. This has as advantage that they are passed in, but might obfuscate the fact that they are unused.

ev-br commented on scipy/linalg/src/_linalg_solve.hh:
```diff
@@ -356,8 +526,7 @@ _solve(PyArrayObject* ap_Am, PyArrayObject *ap_b, T* ret_data, St structure, int
         T* slice_ptr = (T *)(Am_data + (offset/sizeof(T)));
-        copy_slice(scratch, slice_ptr, n, n, strides[ndim-2], strides[ndim-1]); // XXX: make it in one go
-        swap_cf(scratch, data, n, n, n);
+        copy_slice_F(data, slice_ptr, n, n, strides[ndim-2], strides[ndim-1]);
```
Could you please roll back this change? It is unrelated to the banded solver, so let's keep the scope.

We had a discussion with @ilayn several months ago about this part, and IIRC and IIUC @ilayn's statement is that this two-step copy-then-transpose can be faster than a single-step `copy_to_F_order`. My quick-and-dirty testing seemed to indicate that the latter is actually faster, but I did not do any systematic investigation. So let's keep it for a separate PR/issue/discussion (and make sure it includes Ilhan).

ilayn replied:
I did the same measurements two weeks ago for `logm` to fine tune part that and test the two strategies. Not sure what I am doing differently but the double step is still 2x-4x times faster than pushing into fortran order in one go. Especially for nonunity strides I would be really surprised if it is faster and I would be willing to figure it out why.

Pushing it into F-order is what a regular transpose does, with double loop, anyways so I don't think it can beat the tiled recursion version. So maybe C++ missing the `restrict` promises, that C code does, to the compiler or something like that and slowing down on pointer dereferencing. Because I am measuring speedups starting from n >= 10ish until n=>~1500 and after that it does not matter since O(`n**3`) algorithm already dwarfs the transpose.

BasilLiekens replied:
This was mostly done for "uniformity" since the copying strategy differs for the slices of A and b; A uses the "double loop" whereas b does it in one go, looking a bit weird.

However, this is indeed somewhat expanding the scope so I'll roll it back.

ev-br commented on scipy/linalg/src/_linalg_solve.hh:
```diff
@@ -241,13 +281,137 @@ inline void solve_slice_diagonal(
+    npy_intp kl_max = 0;
+    npy_intp ku_max = 0;
+    kls = (npy_intp *)malloc(outer_size * sizeof(npy_intp));
+    kus = (npy_intp *)malloc(outer_size * sizeof(npy_intp));
```
What are these `kls` and `kus` arrays, and why they are `npy_intp` not `CBLAS_INT`?

ev-br reviewed (commented):
Several minor comments below. A larger question is about the data copying strategy here. AFAIU at the moment, you loop over slices, copy each `n-by-n` slice to the scratch area, then extract the bands from this copied area, and only then call LAPACK.   

For an `NxN` matrix with `K` bands, you can avoid an `N*N` sized copy by only copying the band data from the original strided input array?

BasilLiekens committed:
MAINT: address review comments

- Rolled back a change lumping copying and swapping ordering of arrays.
- Moved branching to dedicated function.
- 'to_banded' now directly takes into F-ordered data rather than having to double copy.
- Coalesced memory allocations
- Added 'overwrite_b' into lower-level implementation to appease linter.
- Removed dead code, exception is 'to_banded' function.
- Pruned some auto includes from LSP.
- Made use of 'compute_slice_ptr' function for simplicity.
- Cleaned up some of the comments for clarity.
- Removed redundant '/sizeof(T)''s
- Renamed LAPACK overloads to 'call_xyz'

BasilLiekens commented:
The "copy, then banded" problem has been addressed by directly taking in the data into banded storage by also passing the strides of the array to the `to_banded()` function. 

Comparing the implementation with separate copying, swapping and "to banded" steps to the one that directly translates an input array into banded storage yields the benchmark results under the fold. There is some noise in this benchmark as unaffected parts of the code also show changes. However, as the size of the matrix grows relative to the number of bands, a speedup can be observed as could be expected. (The benchmark has a fixed number of bands.)

<details>
<summary>Benchmark results</summary>

- Separate copying and "to_banded" steps:
  ```
              =============== ============= ============== ============== ============
              --                                              module
              ----------------------------- ------------------------------------------
                   shape        structure    scipy/detect   scipy/assume     numpy
              =============== ============= ============== ============== ============
               (100, 10, 10)       gen         257±1μs        259±3μs       58.8±1μs
               (100, 10, 10)       pos         248±1μs        236±6μs      64.4±0.5μs
               (100, 10, 10)       sym        245±0.8μs       297±5μs      58.8±0.7μs
               (100, 10, 10)     diagonal      24.9±1μs     19.9±0.07μs    58.4±0.6μs
               (100, 10, 10)   tridiagonal     77.5±2μs       74.6±1μs     59.1±0.7μs
               (100, 10, 10)      banded       265±2μs        223±1μs      57.6±0.7μs
               (100, 20, 20)       gen         574±30μs       596±10μs      190±6μs
               (100, 20, 20)       pos         496±10μs       459±8μs       224±1μs
               (100, 20, 20)       sym         504±8μs        757±80μs     189±0.8μs
               (100, 20, 20)     diagonal      60.3±1μs      43.5±0.4μs     186±1μs
               (100, 20, 20)   tridiagonal     173±2μs        162±4μs       187±2μs
               (100, 20, 20)      banded       597±8μs        438±5μs      186±0.8μs
                 (100, 100)        gen         73.2±2μs      73.4±0.5μs    39.4±0.3μs
                 (100, 100)        pos         68.9±1μs       62.4±1μs     40.1±0.9μs
                 (100, 100)        sym        111±0.6μs      78.5±0.7μs    40.4±0.6μs
                 (100, 100)      diagonal     25.3±0.8μs     19.5±0.5μs    39.0±0.3μs
                 (100, 100)    tridiagonal    32.0±0.6μs     26.8±0.5μs    39.7±0.7μs
                 (100, 100)       banded      79.4±0.7μs     41.8±0.2μs    39.3±0.3μs
              =============== ============= ============== ============== ============ 
  ```

- Coalesced calls:
  ```
             =============== ============= ============== ============== =============
             --                                               module
             ----------------------------- -------------------------------------------
                  shape        structure    scipy/detect   scipy/assume      numpy
             =============== ============= ============== ============== =============
              (100, 10, 10)       gen         257±2μs        254±1μs       57.7±0.3μs
              (100, 10, 10)       pos         244±6μs        234±1μs       63.3±0.3μs
              (100, 10, 10)       sym         243±2μs        298±9μs       58.7±0.3μs
              (100, 10, 10)     diagonal     25.1±0.2μs     19.1±0.1μs     57.7±0.4μs
              (100, 10, 10)   tridiagonal     77.2±1μs       71.0±2μs      57.7±0.3μs
              (100, 10, 10)      banded       261±10μs      214±0.9μs      58.0±0.3μs
              (100, 20, 20)       gen         564±8μs        560±5μs        190±4μs
              (100, 20, 20)       pos         471±3μs        442±4μs        224±3μs
              (100, 20, 20)       sym         488±6μs        684±3μs        188±3μs
              (100, 20, 20)     diagonal     62.1±0.6μs     40.3±0.3μs     185±0.5μs
              (100, 20, 20)   tridiagonal     170±2μs        153±4μs        188±2μs
              (100, 20, 20)      banded       591±5μs        411±5μs        192±2μs
                (100, 100)        gen         72.3±1μs       71.8±2μs      40.0±0.2μs
                (100, 100)        pos         69.9±2μs       64.4±1μs      40.7±0.5μs
                (100, 100)        sym        110±0.5μs       79.7±2μs      39.6±0.4μs
                (100, 100)      diagonal     25.2±0.1μs     19.0±0.2μs     39.6±0.7μs
                (100, 100)    tridiagonal    32.0±0.5μs     26.7±0.4μs     39.5±0.2μs
                (100, 100)       banded       79.0±2μs      32.1±0.1μs    39.6±0.09μs
             =============== ============= ============== ============== =============
  ```

- Comparison
  ```
  | Change   | Before [d1e8aece] <solve_speedup~3>   | After [354f2fea] <solve_speedup~2>   |   Ratio | Benchmark (Parameter)                                                             |
  |----------|---------------------------------------|--------------------------------------|---------|-----------------------------------------------------------------------------------|
  | -        | 74.6±1μs                              | 71.0±2μs                             |    0.95 | linalg.BatchedSolveBench.time_solve((100, 10, 10), 'tridiagonal', 'scipy/assume') |
  | -        | 496±10μs                              | 471±3μs                              |    0.95 | linalg.BatchedSolveBench.time_solve((100, 20, 20), 'pos', 'scipy/detect')         |
  | -        | 438±5μs                               | 411±5μs                              |    0.94 | linalg.BatchedSolveBench.time_solve((100, 20, 20), 'banded', 'scipy/assume')      |
  | -        | 596±10μs                              | 560±5μs                              |    0.94 | linalg.BatchedSolveBench.time_solve((100, 20, 20), 'gen', 'scipy/assume')         |
  | -        | 43.5±0.4μs                            | 40.3±0.3μs                           |    0.93 | linalg.BatchedSolveBench.time_solve((100, 20, 20), 'diagonal', 'scipy/assume')    |
  | -        | 41.8±0.2μs                            | 32.1±0.1μs                           |    0.77 | linalg.BatchedSolveBench.time_solve((100, 100), 'banded', 'scipy/assume')         |
  ```
</details>


Similarly, an overload of the `bandwidth()` function has been added that allows to remove the `if(PyArray_IS_F_CONTIGUOUS)` flag that was introduced to deal with the fact that the original `bandwidth()` function assumes that the input is already F-ordered.

ev-br reviewed (commented):
Thanks for the update and sorry for the delay.
This now looks nice, only minor comments left (see a batch of comments below). 

The need for bandwidth detection per slice is unfortunate but probably nothing to do about it given that it's backed in by backwards compat. I hoped to be able to reuse it in interpolate, will need to see what the performance impact is (in there, we do know the bandwidth of all slices). 

Your comment, https://github.com/scipy/scipy/pull/24396#issuecomment-3770161347, hinted that test coverage might use some improvement --- this would be nice indeed, and it could start here (it's not very clear from our comment if the test you're thinking of a test for solve_banded specifically or if it tests `solve(a, b, assume_a="banded")`, too.

I think this can go in after the comments below are addressed, tests are expanded, and the CI is green.

EDIT: Oh, and since this needs a rebase anyway, it probably does not need 25 commits. Please squash them some. Let us know if you need help rebasing.

BasilLiekens commented:
Perhaps something that should be added in the context of #24442; I presume that if that PR lands this one should also be updated accordingly?

However, there is one caveat here: `overwrite_b` can be enabled in a similar fashion to how it was done in that PR. For `overwrite_a` things are little more complicated as it is not guaranteed that the banded storage form will fit into the passed in `a` matrix (This is the case if `2k_l + k_u + 1`, `k_l` and `k_u` being the number of lower and upper bands, respectively, exceeds `N`). Since the buffers can only be allocated after the bandwidth detection step anyways, something like this could be an option after the detection:
```c++
overwrite_a = ldab_max > shape[ndim-2] ? 0 : overwrite_a;
```
Where `ldab_max` is the maximal number of rows for the banded storage format. After this, the same reasoning from #24442 could be applied I suppose.

ev-br reviewed (commented):
Read through once more --- looks good!
There is a small conflict now, could you please rebase on main?
Now that you're rebasing anyway, consider two optional suggestions below. Either way let's get this in once the CI comes back green after rebase.

EDIT: 

> overwrite_b can be enabled in a similar fashion

Yeah, it'd be great. Does not need to hold up this PR though

> For overwrite_a things are little more complicated as it is not guaranteed that the banded storage form will fit into the passed in `a` matrix

Ugh. Definitely fine to keep as a follow-up enhancement.

BasilLiekens committed:
MAINT: addressed some more review comments

ev-br commented:
CI is green, the code LGTM, in it goes. Thank you @BasilLiekens, much appreciated! And congratulations with your first scipy commit. Keep them coming!
